lib/ex_fuzzywuzzy.ex

```diff
@@ -1,18 +1,323 @@
+    apply_ratio(left, right, &do_token_sort_ratio/3, options)
+  end
+
+  @spec do_token_sort_ratio(String.t(), String.t(), ratio_calculator()) :: float()
+  defp do_token_sort_ratio(left, right, ratio_fn) do
+    left
+    |> token_sort_normalizer()
+    |> ratio_fn.(token_sort_normalizer(right))
+  end
+
+  @spec token_sort_normalizer(String.t()) :: String.t()
+  defp token_sort_normalizer(string) do
+    string
+    |> string_normalizer()
+    |> Enum.sort()
+    |> Enum.join(" ")
+  end
+
+  @doc """
+  Like token sort ratio, but a partial ratio - instead of a standard one - is applied
+
+  ```elixir
+  iex> partial_token_sort_ratio("fuzzy wuzzy was a bear", "wuzzy fuzzy was a bear")
+  100.0
+
+  iex> partial_token_sort_ratio("fuzzy was a bear", "fuzzy fuzzy was a bear")
+  90.63
+  ```
+  """
+  @spec partial_token_sort_ratio(String.t(), String.t(), fuzzywuzzy_options()) :: float()
+  def partial_token_sort_ratio(left, right, options \\ []) do
+    apply_ratio(left, right, &do_partial_token_sort_ratio/3, options)
+  end
+
+  @spec do_partial_token_sort_ratio(String.t(), String.t(), ratio_calculator()) :: float()
+  defp do_partial_token_sort_ratio(left, right, _) do
+    do_token_sort_ratio(left, right, fn a, b -> partial_ratio(a, b) / 100 end)
+  end
+
+  @doc """
+  Calculates the token set ratio between two strings, that is the ratio calculated
+  after tokenizing each string, splitting in two sets (a set with fully matching tokens,
+  a set with other tokens), then sorting on set membership and alphabetically
+
+  ```elixir
+  iex> token_set_ratio("fuzzy was a bear", "fuzzy fuzzy was a bear")
+  100.0
+
+  iex> token_set_ratio("fuzzy was a bear", "muzzy wuzzy was a bear")
+  81.58
+  ```
+  """
+  @spec token_set_ratio(String.t(), String.t(), fuzzywuzzy_options()) :: float()
+  def token_set_ratio(left, right, options \\ []), do: apply_ratio(left, right, &do_token_set_ratio/3, options)
+
+  @spec do_token_set_ratio(String.t(), String.t(), ratio_calculator()) :: float()
+  defp do_token_set_ratio(left, right, ratio_fn) do
+    left_tokens = token_set_normalizer(left)
+    right_tokens = token_set_normalizer(right)
+
+    base =
+      left_tokens
+      |> MapSet.intersection(right_tokens)
+      |> Enum.sort()
+      |> Enum.join(" ")
+      |> String.trim()
+
+    left_minus_right = token_set_diff(left_tokens, right_tokens, base)
+
+    right_minus_left = token_set_diff(right_tokens, left_tokens, base)
+
+    [
+      {base, left_minus_right},
+      {base, right_minus_left},
+      {left_minus_right, right_minus_left}
+    ]
+    |> Enum.map(fn {left, right} -> ratio_fn.(left, right) end)
+    |> Enum.max()
+  end
+
+  @spec token_set_normalizer(String.t()) :: MapSet.t()
+  defp token_set_normalizer(string) do
+    string
+    |> string_normalizer()
+    |> MapSet.new()
+  end
+
+  @spec token_set_diff(MapSet.t(), MapSet.t(), String.t()) :: String.t()
+  defp token_set_diff(left, right, prefix) do
```

Comment:
in generale le`(fn ... end).()` mi fanno alzare le antenne, in questo caso si potrebbe scrivere così?

```elixir
body = left
    |> MapSet.difference(right)
    |> Enum.sort()
    |> Enum.join(" ")

prefix <> " " <> body
```

i token dovrebbero essere già trimmati, no?

Reply:
I token sì, ma l'intersezione (prefix) potrebbe essere vuota, per cui `prefix <> " " <> body` può aver bisogno di trimming. Si potrebbe riscrivere come:

```elixir
body = left
    |> MapSet.difference(right)
    |> Enum.sort()
    |> Enum.join(" ")

String.trim(prefix <> " " <> body)
```

che effettivamente guadagna qualcosina in leggibilità